### PR TITLE
feat: Snapshots (without snapshot renaming)

### PR DESCRIPTION
**Pull request by fpena**
## What does this PR do?

This PR implements the state snapshots feature for Reactotron, allowing developers to capture, store, and restore application state snapshots. 

The feature adds a new "Snapshots" tab to the existing State screen, enabling users to take snapshots of their Redux or MobX-State-Tree store state with a single click, view them in an expandable card layout, and restore any captured state back to the connected client.

"Snapshots" enhances the debugging experience by providing state persistence across sessions, enabling developers to capture state at specific points in time and restore it later for testing scenarios or bug reproduction. The implementation includes proper error handling, type safety, and integrates seamlessly with Reactotron's existing WebSocket communication system.

### Key Features:

- 📸 **Snapshot Creation**: One-click state capture from Redux/MobX-State-Tree stores
- 💾 **State Storage**: Local storage with automatic deduplication
- 🔄 **State Restoration**: Restore any snapshot to connected clients
- 📋 **Copy Operations**: Copy individual or all snapshots to clipboard
- 🗑️ **Snapshot Management**: Delete unwanted snapshots
- 🎨 **Modern UI**: Card-based layout with expandable content
- ⏰ **Smart Naming**: Human-readable timestamps (e.g., "Wednesday @ 5:00:15 PM")
- 🛡️ **Error Handling**: Robust validation and error management
- 🔗 **Client Targeting**: Restore to specific clients or active client

### ⚠️ Known Limitations:
- **Snapshot Renaming**: The ability to rename snapshots has not been included due to some modifications the patches within the `patches` folder are making. These patches modify TextInput behavior in React Native macOS, causing issues with text input functionality. This feature will be addressed in a separate PR once the TextInput issues are resolved.

## What GitHub issues does this PR fix?

#41 

### reactotron-macos setup for this PR

1. `gh pr checkout <PR_NUMBER>`
2. `npm install`
3. `npm run pod`
4. `npm run start`
5. `npm run macos`

### reactotron example app setup for this PR

1. `yarn install`
   Add Reactotron standalone server port to example app configuration

```diff
diff --git a/apps/example-app/app/devtools/ReactotronConfig.ts b/apps/example-app/app/devtools/ReactotronConfig.ts
index a5c7d5b6..9b9c1dad 100644
--- a/apps/example-app/app/devtools/ReactotronConfig.ts
+++ b/apps/example-app/app/devtools/ReactotronConfig.ts
@@ -18,6 +18,7 @@ import { Reactotron } from "./ReactotronClient"

 const reactotron = Reactotron.configure({
   name: require("../../package.json").name,
+  port: 9292,
   onConnect: () => {
     /** since this file gets hot reloaded, let's clear the past logs every time we connect */
     Reactotron.clear()

```

2. `yarn workspace example-app start`
3. Start on a different port to avoid conflicting with reactotron-macos
4. Press `i` to launch in iOS simulator
5. Consider `Cmd + Shift + R` to reload reactotron-macos to get connection

#### Prerequisites
* Ensure you have a React Native app connected to Reactotron
* The app should be using Redux, MobX-State-Tree, or another state management library
* Reactotron server should be running and connected

#### Testing Steps

📸 Snapshot Creation
* Navigate to the State screen in Reactotron
* Click on the "Snapshots" tab
* Click the "Create Snapshot" button
* Verify that a new snapshot appears in the list with a timestamp-based name (e.g., "Wednesday @ 5:00:15 PM")
* Test creating multiple snapshots to verify deduplication works

🔍 Snapshot Interaction
* Expand/Collapse: Click on any snapshot card to expand and view the state tree
* Copy Individual: Click the clipboard icon on a snapshot to copy it to clipboard
* Copy All: Click the "Copy All" button to copy all snapshots to clipboard
* Delete: Click the trash icon to delete a snapshot

🔄 State Restoration
* Create a snapshot of your current app state
* Make changes to your app's state (e.g., update Redux store, modify MobX state)
* In Reactotron, click the restore icon (arrow up) on the snapshot
* Verify that your app's state has been restored to the captured state
* Test restoration with different clients if multiple are connected

🎨 UI/UX Testing
* Tab Switching: Switch between "Subscriptions" and "Snapshots" tabs
* Empty State: Test the empty state message when no snapshots exist
* Tooltips: Hover over action buttons to verify tooltips appear

🧪 Edge Cases
* No Active Client: Try creating snapshots when no client is connected
* Invalid State: Test with corrupted or invalid state data
* Multiple Clients: Test snapshot creation/restoration with multiple connected clients
* Large State: Test with large state objects to verify performance

🔧 Technical Validation
* Deduplication: Create snapshots rapidly to test duplicate prevention
* Memory Management: Create and delete many snapshots to test cleanup
* Error Handling: Test with network disconnections during snapshot operations

Expected Results
✅ Snapshots are created with proper timestamps
✅ State restoration works correctly
✅ Copy operations work (check clipboard)
✅ UI is responsive and intuitive
✅ Error handling works for edge cases
✅ No memory leaks with repeated operations

# Screenshot/Video of "How to test this PR?"

<video src='https://github.com/user-attachments/assets/5f8dddfc-2c32-473c-b13f-f14715e0e0ba' width=180/>

### CODE REVIEW

File: app/screens/StateScreen.tsx [hunk 1]
```diff
@@ -39,6 +50,72 @@ export function StateScreen() {
     }))
   }
 
+  const createSnapshot = () => {
+    if (!activeTab) {
+      console.log("No active client to create snapshot from")
+      return
+    }
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: Prefer one-liner guards if possible.

```suggestion
    if (!activeTab) return console.log("No active client to create snapshot from")
```

File: app/screens/StateScreen.tsx [hunk 2]
```diff
@@ -39,6 +50,72 @@ export function StateScreen() {
...
+    sendToCore("state.backup.request", { clientId: activeTab })
+  }
+
+  const copySnapshotToClipboard = (snapshot: Snapshot) => {
+    try {
+      IRClipboard.setString(JSON.stringify(snapshot.state, null, 2))
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: Should we be formatting the JSON string like this, or just stringifying without formatting?

Comment (jamonholmgren) [reply app/screens/StateScreen.tsx]: I think this should be exported to a utility function in another file rather than putting it inline in the component body. It doesn't depend on any state except what's passed in.

File: app/screens/StateScreen.tsx [hunk 3]
```diff
@@ -39,6 +50,72 @@ export function StateScreen() {
...
+
+  const copyAllSnapshotsToClipboard = () => {
+    try {
+      console.log("Copying all snapshots to clipboard", snapshots)
+      IRClipboard.setString(JSON.stringify(snapshots, null, 2))
+      console.log("All snapshots copied to clipboard")
+    } catch (error) {
+      console.error("Failed to copy snapshots to clipboard:", error)
+    }
+  }
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: This can be extracted to a utility function as well. You can either pass in the `snapshots` to the function or just get them on-demand with `withGlobal` (or if @SeanBarker182 has renamed that function, whatever he landed on).

File: app/screens/StateScreen.tsx [hunk 4]
```diff
@@ -39,6 +50,72 @@ export function StateScreen() {
...
+      }
+      return newSet
+    })
+  }
+
+  const restoreSnapshot = (snapshot: Snapshot) => {
+    if (!snapshot || !snapshot.state) {
+      console.error("Invalid snapshot: missing state data")
+      return
+    }
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: single-line preferred here

Comment (jamonholmgren) [reply app/screens/StateScreen.tsx]: and below in the targetClientId check too

File: app/screens/StateScreen.tsx [hunk 5]
```diff
@@ -39,6 +50,72 @@ export function StateScreen() {
...
+
+    // Send the restore command to the client
+    sendToCore("state.restore.request", {
+      clientId: targetClientId,
+      state: snapshot.state,
+    })
+
+    console.log(`Restoring snapshot "${snapshot.name}" to client ${targetClientId}`)
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: Should these console.log lines be removed for this PR?

File: app/screens/StateScreen.tsx [hunk 6]
```diff
@@ -49,55 +126,161 @@ export function StateScreen() {
...
-                ...prev,
-                [activeTab]: [],
-              }))
-              sendToCore("state.values.subscribe", { paths: [], clientId: activeTab })
-              setActiveTab("")
-            }}
-          >
-            <Text>Clear State</Text>
-          </Pressable>
+    <>
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: Why is the ScrollView wrapped in a Fragment, here? Looks like it's just one element.

File: app/screens/StateScreen.tsx [hunk 7]
```diff
@@ -49,55 +126,161 @@ export function StateScreen() {
...
-            {clientStateSubscriptions.map((subscription, index) => (
-              <View key={`${subscription.path}-${index}`} style={$stateItemContainer()}>
-                <Text style={$pathText()}>
-                  {subscription.path ? subscription.path : "Full State"}
+        <View style={$tabsContainer()}>
+          <Tab id="subscriptions" label="Subscriptions" tabgroup="activeStateTab" />
+          <Tab id="snapshots" label="Snapshots" tabgroup="activeStateTab" />
+        </View>
+        <View style={$stateContainer()}>
+          {activeStateTab === "Subscriptions" ? (
```
Comment (jamonholmgren) [inline app/screens/StateScreen.tsx]: I think we're at the point where this should be refactored into two sub components. Likely something like `StateSubscriptions` and `StateSnapshots`.

File: app/state/connectToServer.ts [hunk 8]
```diff
@@ -188,6 +189,50 @@ export function connectToServer(props: { port: number } = { port: 9292 }): Unsub
         setCustomCommands((prev) => prev.filter((cmd) => cmd.id !== commandId))
         return
       }
+
+      // Handle state backup response
+      if (data.cmd.type === "state.backup.response") {
+        console.log("Received state.backup.response:", data.cmd)
```
Comment (jamonholmgren) [inline app/state/connectToServer.ts]: probably remove console.logs?

File: app/state/connectToServer.ts [hunk 9]
```diff
@@ -188,6 +189,50 @@ export function connectToServer(props: { port: number } = { port: 9292 }): Unsub
         setCustomCommands((prev) => prev.filter((cmd) => cmd.id !== commandId))
         return
       }
+
+      // Handle state backup response
+      if (data.cmd.type === "state.backup.response") {
```
Comment (jamonholmgren) [inline app/state/connectToServer.ts]: This should be a `CommandType`

**Review by jamonholmgren — CHANGES_REQUESTED** [review]
Some things to address, but headed the right direction

**Comment by jamonholmgren** [comment]
Add screenshots of the feature too, please

File: app/screens/StateScreen.tsx (hunk 2 above)
Comment (fpena) [reply app/screens/StateScreen.tsx]: @jamonholmgren 

> Should we be formatting the JSON string like this, or just stringifying without formatting?

Is your concern here about performance and/or storage usage?

Comment (jamonholmgren) [reply app/screens/StateScreen.tsx]: > Is your concern here about performance and/or storage usage?

I'm mostly thinking about storage. Perf could potentially be an issue too. But I'm okay if you have a good reason to format it, just asking the question since I wasn't sure.

Comment (fpena) [reply app/screens/StateScreen.tsx]: This is for copying the state to the clipboard, so I would not say there's a storage issue. 
Also, I believe I don't have a good reason to format the JSON, so I'm removing it.

Comment (jamonholmgren) [reply app/screens/StateScreen.tsx]: You know, actually, since it's supposed to go in the clipboard, let's keep the formatting. I think you were right the first time. They might paste somewhere where formatting matters.

File: app/components/State/StateSnapshots.tsx [hunk 10]
```diff
@@ -0,0 +1,183 @@
...
+import { themed, useTheme, useThemeName } from "../../theme/theme"
+import { TreeViewWithProvider } from "../TreeView"
+import { Divider } from "../Divider"
+import { Icon } from "../Icon"
+import { Tooltip } from "../Tooltip"
+import { useState } from "react"
+import { useGlobal } from "../../state/useGlobal"
+import { sendToCore } from "../../state/connectToServer"
+import IRClipboard from "../../native/IRClipboard/NativeIRClipboard"
+import type { Snapshot } from "app/types"
```
Comment (jamonholmgren) [inline app/components/State/StateSnapshots.tsx]: My editor will sometimes import things from `app/...` but I've had issues with that working. I'd probably change it to `../../types` to stay consistent with the above.

File: app/components/State/StateSnapshots.tsx [hunk 11]
```diff
@@ -0,0 +1,183 @@
...
+
+export function StateSnapshots() {
+  const theme = useTheme()
+  const [themeName] = useThemeName()
+  const [snapshots, setSnapshots] = useGlobal<Snapshot[]>("snapshots", [])
```
Comment (jamonholmgren) [inline app/components/State/StateSnapshots.tsx]: This isn't an error on your part, but I think this hook could be improved by making a custom `useSnapshot` hook. I would export an object with the following:

```tsx
const { snapshots, setSnapshots, restoreSnapshot, deleteSnapshot } = useSnapshots()
```

Perhaps even an `addSnapshot` as well.

Use this custom hook anywhere you're using snapshots.
